netket/optimizer/sr/s_jacobian_mat.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Callable, Optional, Union, Tuple, Any
+from functools import partial
+
+import jax
+from jax import numpy as jnp
+from flax import struct
+
+from netket.utils.types import PyTree
+import netket.jax as nkjax
+
+from .base import AbstractSMatrix
+
+
+@struct.dataclass
+class JacobianSMatrix(AbstractSMatrix):
+    """
+    Semi-lazy representation of an S Matrix behaving like a linear operator.
+
+    The matrix of gradients O is computed on initialisation, but not S, 
+    which can be computed by calling :code:`to_dense`.
+    The details on how the ⟨S⟩⁻¹⟨F⟩ system is solved are contaianed in
+    the field `sr`.
+    """
+
+    sr: SRJacobian
+    """Parameters for the solution of the system."""
+
+    O: jnp.ndarray
+    """Gradients O_ij = ∂log ψ(σ_i)/∂p_j of the neural network 
+    for all samples σ_i at given values of the parameters p_j
+    Average <O_j> subtracted for each parameter
+    Divided through by sqrt(#samples) to normalise S matrix
+    If scale is not None, normalised to unit magnitude"""
+
+    scale: Optional[jnp.ndarray] = None
+    """If not None, gives scale factors with which O is normalised"""
+
+    x0: Optional[PyTree] = None
+    """Optional initial guess for the iterative solution."""
+
+    @jax.jit
+    def __matmul__(self, vec: Union[PyTree, jnp.ndarray]) -> Union[PyTree, jnp.ndarray]:
+        if not hasattr(vec, "ndim"):
+            vec, unravel = nkjax.tree_ravel(vec)
+        else:
+            unravel = lambda x: x
+
+        return unravel(
```

Comment:
You should only unravel the output if you ravelled the input. 
Otherwise this won't work with scipy solvers.
Check the implementation of onthefly.

Reply:
In that case, `unravel` is just an identity function (see line 60), but I can change it if confusing

Reply:
ah, yes, sorry.
It's a bit confusing but might be ok